python/ray/metrics_agent.py

```diff
@@ -0,0 +1,180 @@
+
+        # OpenCensus classes.
+        self.stats = stats_module.stats
+        self.view_manager = self.stats.view_manager
+        self.stats_recorder = self.stats.stats_recorder
+        # Port where we will expose metrics.
+        self.metrics_export_port = metrics_export_port
+        # Measure to record metrics.
+        self._measure_map = self.stats_recorder.new_measurement_map()
+        # metric name(str) -> view (view.View)
+        # It is updated dynamically as new metrics
+        # are delivered from cpp processes.
+        self._registry = defaultdict(lambda: None)
+        # Lock required because gRPC server uses multi-threads.
+        self._lock = threading.Lock()
+        # Whether or not there are metrics that are missing description and
+        # units information. This is used to dynamically update registry.
+        self._missing_information = False
+
+        # Configure exporter. (We currently only support prometheus).
+        self.view_manager.register_exporter(
+            prometheus_exporter.new_stats_exporter(
+                prometheus_exporter.Options(
+                    namespace="ray", port=metrics_export_port)))
+
+    @property
+    def registry(self):
+        return self._registry
+
+    def record_metrics_points(self, metrics_points):
+        with self._lock:
+            for metric_point in metrics_points:
+                self._register_if_needed(metric_point)
+                self._record(metric_point)
+            return self._missing_information
+
+    def _record(self, metric_point):
+        """Record a single metric point to export.
+
+        NOTE: When this method is called, the caller should acquire a lock.
+
+        Args:
+            metric_point metric point defined in common.proto
+        """
+        metric_name = metric_point.metric_name
+        tags = metric_point.tags
+
+        metric = self._registry.get(metric_name)
+        # Metrics should be always registered dynamically.
+        assert metric
+
+        tag_map = tag_map_module.TagMap()
+        for key, value in tags.items():
+            tag_key = tag_key_module.TagKey(key)
+            tag_value = tag_value_module.TagValue(value)
+            tag_map.insert(tag_key, tag_value)
+
+        metric_value = metric_point.value
+        # The class should be thread-safe because GRPC server
+        # uses it with a threadpool.
+        self._measure_map.measure_float_put(metric.measure, metric_value)
+        # NOTE: When we record this metric, timestamp will be renewed.
+        self._measure_map.record(tag_map)
+
+    def _register_if_needed(self, metric_point):
+        """Register metrics if they are not registered.
+
+        NOTE: When this method is called, the caller should acquire a lock.
+
+        Unseen metrics:
+            Register it with Gauge type metrics. Note that all metrics in
+            the agent will be gauge because sampling is already done
+            within cpp processes.
```

Comment:
This seems a little bit strange to me - does this affect anything downstream? e.g., will prometheus aggregate them differently? or is it purely vanity here?

Reply:
I meant that counter, histogram, or sum doesn't make sense to be here because reporter just collects momental values (gauge) that are already counted or sampled (histogram for example), or summed inside cpp processes. 

afaik, Prometheus aggregates them differently in a client level.

Reply:
Modified comments. Please check it out!